### PR TITLE
[pulseaudio] source: use thread safe collection and force reconnection

### PR DESCRIPTION
**Pull request by GiviMAD**
Signed-off-by: Miguel Álvarez Díez <miguelwork92@gmail.com>

Use thread safe collection for storing the source pipe output references.
Also added a forced disconnection after the 3 consecutive IOExceptions while reading the socket input to force a reconnection.
Possible fix for https://github.com/openhab/openhab-addons/issues/12432.
 
@lolodomo can you try to reproduce the error with this version?

### CODE REVIEW

**Comment by lolodomo** [comment]
I tested only with the "listen and answer" feature so with no risk of concurrency access. 
But ok I am going to test with that version.

**Comment by GiviMAD** [comment]
> I tested only with the "listen and answer" feature so with no risk of concurrency access.
> But ok I am going to test with that version.

I think the problem was that as the list was not thread safe the task for the writing to the pipe outputs was writing to one already closed, so the error could also happen using just the "listen and answer" mode. Hope I'm right and it gets solved. Let me know and thanks for the testing.

**Comment by GiviMAD** [comment]
Also the forced disconnection after the 3 consecutive IOExceptions while reading, solved a problem I was facing, the stream was not able to recover after I a system sleep of openHAB server (during my testing this is my mac). 
I been connected to the speaker since I summited the pr, like 5 hours, and the dialog keeps working nicely :)

**Comment by GiviMAD** [comment]
@lolodomo does it work for you?

**Comment by lolodomo** [comment]
Not yet tried, maybe this evening.

**Comment by lolodomo** [comment]
I just tried. It failed at the end of my first call and the second call fully failed.
Here are my logs:
```
20:19:11.452 [DEBUG] [.core.voice.internal.VoiceManagerImpl] - Executing a simple dialog for source Entrée audio Sony Eye (pulseaudio:source:rpi2:entree)
20:19:13.873 [DEBUG] [b.core.voice.internal.DialogProcessor] - RecognitionStartEvent event received
20:19:21.590 [DEBUG] [b.core.voice.internal.DialogProcessor] - RecognitionStopEvent event received
20:19:21.596 [DEBUG] [b.core.voice.internal.DialogProcessor] - SpeechRecognitionEvent event received
20:19:21.599 [DEBUG] [b.core.voice.internal.DialogProcessor] - Text recognized: allume les spots
20:19:21.604 [DEBUG] [b.core.voice.internal.DialogProcessor] - Interpretation exception: Désolé, je n'ai pas compris.
20:19:27.967 [WARN ] [eaudio.internal.PulseAudioAudioSource] - IOException while writing to from pulse source pipe: null
...
20:19:57.735 [DEBUG] [.core.voice.internal.VoiceManagerImpl] - Executing a simple dialog for source Entrée audio Sony Eye (pulseaudio:source:rpi2:entree)
20:19:57.742 [WARN ] [eaudio.internal.PulseAudioAudioSource] - IOException while writing to from pulse source pipe: Pipe not connected
20:19:58.508 [DEBUG] [b.core.voice.internal.DialogProcessor] - RecognitionStartEvent event received
20:20:04.605 [DEBUG] [b.core.voice.internal.DialogProcessor] - RecognitionStopEvent event received
20:20:04.612 [DEBUG] [b.core.voice.internal.DialogProcessor] - SpeechRecognitionErrorEvent event received
```
The time to write this text, I tried again and it did not get errors this time.
It failed again at the 5th or 6th call.
So it continues to fail randomly, around one every 3 calls.
Sorry but I see no progress.

**Comment by GiviMAD** [comment]
No idea what can be happening, I never get that error. I will try the "listen and answer" once it merged, maybe something is different there. Can you check if you face that error using the continuos dialog execution?

For me this pr still solves the reconnection problem so I would like to merge it any way.

**Comment by GiviMAD** [comment]
Ok I think I understand the problem I will make another try.

**Comment by GiviMAD** [comment]
Fixed, the "Pipe not connected" should be solved. Sorry I didn't realize before.

**Comment by GiviMAD** [comment]
@lolodomo can you give it a try?

Also, should I try to print the exception cause message to avoid this logs?
```
20:19:27.967 [WARN ] [eaudio.internal.PulseAudioAudioSource] - IOException while writing to from pulse source pipe: null
```
Because I think that if the message is null there is probably there is an underling exception. WDYT?

**Comment by lolodomo** [comment]
Sorry, it is not better with your last version.
I am stopping tests for today.

**Comment by GiviMAD** [comment]
> As a review note for myself, I will have to understand what is this famous "Thread.sleep(0);" you try sometimes to use.

It gives priority to other threads.
What I'm trying to do there is: as one thread is iterating the collection of pipe outputs, to avoid to close that output while a write operation is happening, I remove the output ref from the collection then call sleep(0) and then I close the output.

If you have time to continue testing this one in the afternoon please send me another output using the latest build. I would like to achieve here the best solution we can, as I will use the same approximation to solve the problem with the javasound source on windows, and I hope to be able to do it before the next release.
Thanks for the time, and if you have some idea about what can be the error source please let me know.

**Comment by lolodomo** [comment]
Your last version does not compile. First you need to apply spotless. Even after that, I got these errors:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project org.openhab.binding.pulseaudio: Compilation failure: Compilation failure:
[ERROR] D:\dev\openhab3\git\openhab-addons\bundles\org.openhab.binding.pulseaudio\src\main\java\org\openhab\binding\pulseaudio\internal\PulseAudioAudioSource.java:[157,45] The method getExceptionMessage(java.io.IOException) is undefined for the type org.openhab.binding.pulseaudio.internal.PulseAudioAudioSource
[ERROR] D:\dev\openhab3\git\openhab-addons\bundles\org.openhab.binding.pulseaudio\src\main\java\org\openhab\binding\pulseaudio\internal\PulseAudioAudioSource.java:[160,45] The method getExceptionMessage(java.lang.RuntimeException) is undefined for the type org.openhab.binding.pulseaudio.internal.PulseAudioAudioSource
[ERROR] D:\dev\openhab3\git\openhab-addons\bundles\org.openhab.binding.pulseaudio\src\main\java\org\openhab\binding\pulseaudio\internal\PulseAudioAudioSource.java:[164,92] The method getExceptionMessage(java.io.IOException) is undefined for the type org.openhab.binding.pulseaudio.internal.PulseAudioAudioSource
[ERROR] D:\dev\openhab3\git\openhab-addons\bundles\org.openhab.binding.pulseaudio\src\main\java\org\openhab\binding\pulseaudio\internal\PulseAudioAudioSource.java:[172,97] The method getExceptionMessage(java.lang.RuntimeException) is undefined for the type org.openhab.binding.pulseaudio.internal.PulseAudioAudioSource
[ERROR] D:\dev\openhab3\git\openhab-addons\bundles\org.openhab.binding.pulseaudio\src\main\java\org\openhab\binding\pulseaudio\internal\PulseAudioAudioSource.java:[203,13] Return type for the method is missing
[ERROR] D:\dev\openhab3\git\openhab-addons\bundles\org.openhab.binding.pulseaudio\src\main\java\org\openhab\binding\pulseaudio\internal\PulseAudioAudioSource.java:[203,13] Return type for the method is missing
```

**Comment by lolodomo** [comment]
I solved the compilation error with these changes:
```
    private @Nullable String getExceptionMessage(Exception e) {
        String message = e.getMessage();
        var cause = e.getCause();
        if (message == null && cause != null) {
            message = cause.getMessage();
        }
        return message;
    }
```

**Comment by GiviMAD** [comment]
Fixed, didn't check that last one correctly, sorry.

**Comment by lolodomo** [comment]
Last tests.
I have now less warnings.
When the command succeeds, I have a final IOException (it looks like it is systematic when I have a detection - see after), like this:
```
22:45:19.567 [DEBUG] [.core.voice.internal.VoiceManagerImpl] - Executing a simple dialog for source Entrée audio Sony Eye (pulseaudio:source:rpi2:entree)
22:45:19.571 [DEBUG] [ternal.PulseaudioSimpleProtocolStream] - Simple TCP Stream connecting
22:45:19.601 [DEBUG] [ternal.PulseaudioSimpleProtocolStream] - Scheduling disconnect
22:45:20.372 [DEBUG] [b.core.voice.internal.DialogProcessor] - RecognitionStartEvent event received
22:45:20.377 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:20.382 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:20.386 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:22.021 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:22.027 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:22.102 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:22.106 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:22.179 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:22.185 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:22.189 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:24.763 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:25.452 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:25.460 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:25.469 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:45:26.715 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume"
}
22:45:26.984 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la"
}
22:45:26.996 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la"
}
22:45:27.002 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la"
}
22:45:27.419 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:27.426 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:27.717 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:27.726 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:28.111 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:28.117 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:28.423 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:28.430 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:28.437 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière"
}
22:45:28.547 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.555 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.661 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.670 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.745 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.751 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.826 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.834 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.840 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : "allume la lumière chambre"
}
22:45:28.942 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Result: {
  "text" : "allume la lumière chambre"
}
22:45:28.946 [DEBUG] [b.core.voice.internal.DialogProcessor] - RecognitionStopEvent event received
22:45:28.950 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Final: allume la lumière chambre
22:45:28.954 [DEBUG] [b.core.voice.internal.DialogProcessor] - SpeechRecognitionEvent event received
22:45:28.958 [DEBUG] [b.core.voice.internal.DialogProcessor] - Text recognized: allume la lumière chambre
22:45:28.965 [DEBUG] [b.core.voice.internal.DialogProcessor] - Interpretation result: Ok.
22:45:28.981 [INFO ] [openhab.event.ItemCommandEvent       ] - Item 'LampesHueChambre' received command ON
22:45:29.008 [DEBUG] [seaudio.internal.ConvertedInputStream] - Duration of input stream : 1535
22:45:29.013 [DEBUG] [seaudio.internal.ConvertedInputStream] - Sound is a MP3. Trying to reencode it
22:45:29.034 [DEBUG] [seaudio.internal.ConvertedInputStream] - Sound is not in the target format. Trying to reencode it
22:45:29.039 [DEBUG] [ternal.PulseaudioSimpleProtocolStream] - Simple TCP Stream connecting
22:45:29.135 [DEBUG] [lseaudio.internal.PulseAudioAudioSink] - Sleep time to let the system play sound : 1449
22:45:30.589 [DEBUG] [ternal.PulseaudioSimpleProtocolStream] - Scheduling disconnect
22:45:30.637 [DEBUG] [ternal.PulseaudioSimpleProtocolStream] - Scheduling disconnect
22:45:30.643 [WARN ] [eaudio.internal.PulseAudioAudioSource] - IOException while writing to from pulse source pipe: null
```

But now, I encounter a new problem, often, Vosk is just detecting nothing, like for example:
```
22:59:05.079 [DEBUG] [.core.voice.internal.VoiceManagerImpl] - Executing a simple dialog for source Entrée audio Sony Eye (pulseaudio:source:rpi2:entree)
22:59:05.089 [DEBUG] [ternal.PulseaudioSimpleProtocolStream] - Scheduling disconnect
22:59:05.837 [DEBUG] [b.core.voice.internal.DialogProcessor] - RecognitionStartEvent event received
22:59:05.845 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:05.850 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:05.853 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:05.857 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:07.432 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:07.438 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.330 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.338 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.345 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.495 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.503 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.619 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.625 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.633 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.708 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.714 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.720 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.818 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.834 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:10.998 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.005 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.013 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.087 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.094 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.101 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.235 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.250 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.351 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.357 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.365 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.460 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.466 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.471 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.607 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.629 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.727 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.740 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.817 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.823 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.830 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.949 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:11.955 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.052 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.059 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.174 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.181 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.368 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.374 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.382 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.487 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.493 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Partial: {
  "partial" : ""
}
22:59:12.646 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Result: {
  "text" : ""
}
22:59:12.650 [DEBUG] [b.core.voice.internal.DialogProcessor] - RecognitionStopEvent event received
22:59:12.654 [DEBUG] [voice.voskstt.internal.VoskSTTService] - Final:
22:59:12.657 [DEBUG] [b.core.voice.internal.DialogProcessor] - SpeechRecognitionErrorEvent event received
22:59:12.695 [DEBUG] [seaudio.internal.ConvertedInputStream] - Duration of input stream : 6963
22:59:12.698 [DEBUG] [seaudio.internal.ConvertedInputStream] - Sound is a MP3. Trying to reencode it
22:59:12.706 [DEBUG] [seaudio.internal.ConvertedInputStream] - Sound is not in the target format. Trying to reencode it
```
Note that in this case, I have no final IOException.

**Comment by lolodomo** [comment]
Just an idea: you are not checking the value of lengthRead. Maybe it can be -1 at the end?

**Comment by lolodomo** [comment]
> Just an idea: you are not checking the value of lengthReaf?
> I will trace this value, maybe it can help to understand.

Tested, the value is 1024 when the exception occur:
```
07:47:12.490 [WARN ] [eaudio.internal.PulseAudioAudioSource] - IOException while writing to from pulse source pipe (lengthRead = 1024): Pipe closed
```
The exception is triggered by this code:
```
                                    output.write(buffer, 0, lengthRead);
                                    if (pipeOutputs.contains(output)) {
                                        output.flush();
                                    }
```
I am going to check if this is the write or the flush...

**Comment by lolodomo** [comment]
Here is with the stack trace:
```

07:59:03.868 [WARN ] [eaudio.internal.PulseAudioAudioSource] - IOException while writing to from pulse source pipe (lengthRead = 1024): null
java.io.InterruptedIOException: null
        at java.io.PipedInputStream.awaitSpace(PipedInputStream.java:275) ~[?:?]
        at java.io.PipedInputStream.receive(PipedInputStream.java:231) ~[?:?]
        at java.io.PipedOutputStream.write(PipedOutputStream.java:149) ~[?:?]
        at org.openhab.binding.pulseaudio.internal.PulseAudioAudioSource.lambda$1(PulseAudioAudioSource.java:151) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```
So it is the write call.

**Comment by lolodomo** [comment]
Ok, I think I am understanding the problem:
```

08:20:34.353 [WARN ] [eaudio.internal.PulseAudioAudioSource] - Closing PipedInputStream
08:20:34.360 [WARN ] [eaudio.internal.PulseAudioAudioSource] - unregisterPipe
08:20:34.367 [WARN ] [eaudio.internal.PulseAudioAudioSource] - IOException while writing to from pulse source pipe (lengthRead = 1024): null
java.io.InterruptedIOException: null
        at java.io.PipedInputStream.awaitSpace(PipedInputStream.java:275) ~[?:?]
        at java.io.PipedInputStream.receive(PipedInputStream.java:231) ~[?:?]
        at java.io.PipedOutputStream.write(PipedOutputStream.java:149) ~[?:?]
        at org.openhab.binding.pulseaudio.internal.PulseAudioAudioSource.lambda$1(PulseAudioAudioSource.java:152) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```
The input stream is closed and you run unregisterPipe inside this close. Then the output is probably used by the task while it is removed.

**Comment by GiviMAD** [comment]
> The input stream is closed and you run unregisterPipe inside this close. Then the output is probably used by the task while it is already removed.

Great, thanks for the investigation! 
So I can remove the Thread.sleep as it's not having the desired effect.

> Maybe stopPipeWriteTask should be called before closing the pipe input stream ?

Yes, but as the write task is shared for all the active pipes, we should also ignore that error, because when multiple pipes are active we can not stop the task (the stopPipeWriteTask checks that the collection of pipe outputs is empty before stopping the task). I'll add a debug log in that case instead of a warning, so we can just saw the interrupted exception in the debug logs.

**Comment by GiviMAD** [comment]
Seems like I was wrong. 

I was obfuscated thinking that the error was thrown when the socket is closed but it is raised when calling stopPipeWriteTask after removing the last pipe from the collection.

I could correctly replicate the error just starting and stoping the dialogProcessor, and moving the stopPipeWriteTask after the sleep(0) I stop facing it. So the sleep was right just in the incorrect place.

Can you verify that the error is gone for you also with the latests changes?

**Comment by GiviMAD** [comment]
Still can produce the error eventually by executing this on the console 'voice startdialog && voice stopdialog'. The call to sleep(0) mitigates the error but doesn't solved it completely.
So I have added an if clause to prevent this warning just in case.

File: bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java [hunk 1]
```diff
@@ -133,24 +135,46 @@ private synchronized void registerPipe(PipedOutputStream pipeOutput) {
...
             this.pipeWriteTask = executor.submit(() -> {
                 int lengthRead;
                 byte[] buffer = new byte[1024];
+                int readRetries = 3;
                 while (!pipeOutputs.isEmpty()) {
                     var stream = getSourceInputStream();
                     if (stream != null) {
                         try {
                             lengthRead = stream.read(buffer);
+                            readRetries = 3;
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java:149]: Is a retry mechanism really required? 
During my previous tests  I never entered in that case. Did you? 
If required, is it a workaround for a bad implementation?

File: bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java [hunk 2]
```diff
@@ -133,24 +135,46 @@ private synchronized void registerPipe(PipedOutputStream pipeOutput) {
...
                             for (var output : pipeOutputs) {
-                                output.write(buffer, 0, lengthRead);
-                                output.flush();
+                                try {
+                                    output.write(buffer, 0, lengthRead);
+                                    if (pipeOutputs.contains(output)) {
+                                        output.flush();
+                                    }
+                                } catch (IOException e) {
+                                    if (e instanceof InterruptedIOException && pipeOutputs.isEmpty()) {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java:157]: Is there a hierarchy link between InterruptedIOException and IOException? 
Why not catching InterruptedIOException directly?

Comment (lolodomo) [inline bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java:157]: Is there a hierarchy link between InterruptedIOException and IOException? 
Why not catching InterruptedIOException directly?

**Review by lolodomo — CHANGES_REQUESTED** [review]

**Comment by lolodomo** [comment]
I will probably not have the time to test your last version before next week.

File: bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java:157]: Verified. No link. 
Your condition will never be true.

File: bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java (hunk 1 above)
Comment (GiviMAD) [reply bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java:149]: I realize that after a system sleep on the openHAB server the socket doesn't report to be closed but you can't read more from it (it gives a read timeout exception on all the read calls). I think there is some cases where the socket just hangs and became unusable but the jvm is not aware of it.
I saw a similar retry mechanism somewhere else in this binding where the socket disconnection is forced after 3 consecutive IOException so I made the same here and it solves the problem.

> If required, is it a workaround for a bad implementation?

So I think the answer is no, or at least not a bad implementation inside the binding.

**Comment by lolodomo** [comment]
New tests done. Half of my tries are failing with SpeechRecognitionErrorEvent due to STT service returning an empty string. I would not be surprised that the problem is in fact in the audio source management (even if now there is no more WARN log) and not in the Vosk STT service.
Note that the new retry mechanism was never triggered during my tests.

IMHO, there is really something wrong., either in Pulseaudio or in Vosk STT.

That being said, if you are sure of you and if this PR is really an improvement for you, I am ok to merge it.

**Comment by GiviMAD** [comment]
Yes please, I would like to merge it, a couple of thinks are wrong without this.

Let me know if you are able to isolate the problem in one service. Maybe you can try other combinations, like PulseAudio + Watson STT or JavaSound + VoskSTT.

**Pull request merged**